### PR TITLE
add client-side router

### PR DESCRIPTION
**Pull request by smacker**
Ref: #399

### CODE REVIEW

**Comment by carlosms** [comment]
link to issue: #399

**Comment by carlosms** [comment]
Continuing the conversation from #425.
The code looks good and not too complex. And it opens the door for a future UI design with proper routes like `/org-name/settings`, `/org-name/repo-name/settings`.

But still, it adds bloat that I don't know if we want to add at this stage (basic UI was the key requirement). Besides losing the nice `/login` and `/logout` routes in the address bar, are there any downsides to not use the router, @smacker? How would the auth management work?

**Comment by smacker** [comment]
There are no technical limitations to do everything without a router. It works just fine on the previous PR. Most probably it would require only new logout method of App component and pass it down as props to every component that need to do logout (including failed api calls). And of course there are some more hacky solutions like just `Token.delete() && window.reload()`.

I see this advantages of adding the router on the current stage (from the biggest to the smallest in my option):
- It might be much of a headache to add a router to an existing project that didn't have it before. Due to synchronization of the URL and internal state.
- A router naturally introduces abstraction of pages and separation of them from components. I find it useful while building web applications.
- Nice urls

And the disadvantages:
- it requires a bit more code due to better structure / new page & route abstractions

As a conclusion maybe I wouldn't implement it right now if there were no requirements for clean url. But now when it's already here I see more benefits than harm by merging it.

File: frontend/src/services/auth.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+export interface User {
+  name: string;
+}
+
+const localStorageKey = 'token';
+
+class AuthService {
+  private _user: User | null = null;
+
+  get token() {
```
Comment (carlosms) [inline frontend/src/services/auth.ts]: should the getters in this class have return type?

Comment (carlosms) [reply frontend/src/services/auth.ts]: I'm not sure right now. I guess if it's good enough for the `strict` setting in `tsconfig` we can keep using the inferred return; we can always add them later if we see we prefer the explicit types.

**Pull request merged**